### PR TITLE
C-Haines geojson

### PR DESCRIPTION
**Pull request by Sybrand**
Test here: https://wps-pr-1038.apps.silver.devops.gov.bc.ca/c-haines

- removed write to db entirely
- transforming geojson then uploading to object store
- geojson migration
- some changes to kml migration and generation
- introduce async library for s3
- unit tests

NEXT PR:
- get rid of minio (so everything can be async).
- remove migration scripts.
- remove database tables.
- revert linting exclusion for duplicates.

### CODE REVIEW

File: api/app/c_haines/fetch.py [hunk 1]
```diff
@@ -99,22 +91,74 @@ def fetch_network_link_kml() -> str:
...
-                result.model_runs.append(model_run_predictions)
-                prev_model_run_id = model_run_id
-            model_run_predictions.prediction_timestamps.append(prediction_timestamp)
+def extract_model_run_prediction_from_path(prediction_path: str) -> Tuple[str, datetime, datetime]:
+    """ Extract model abbreviation, model run timestamp and prediction timestamp from prediction path """
+    name_split = prediction_path.strip('/').split('/')
+
+    prediction_timestamp = datetime.fromisoformat(name_split[-1].rstrip('.json')+'+00:00')
+
+    hour = int(name_split[-2])
```
Comment (andrea-williams) [inline api/app/c_haines/fetch.py:100]: isn't this functionality captured in `extract_model_run_timestamp_from_path()` method immediately below?

File: api/app/c_haines/fetch.py [hunk 2]
```diff
@@ -99,22 +91,74 @@ def fetch_network_link_kml() -> str:
...
+                prev_model_run_timestamp = None
+                for prediction in prediction_result['Contents']:
+                    model, model_run_timestamp, prediction_timestamp = extract_model_run_prediction_from_path(
+                        prediction['Key'])
+                    if prev_model_run_timestamp != model_run_timestamp:
+                        prev_model_run_timestamp = model_run_timestamp
+                        model_run_predictions = CHainesModelRunPredictions(
+                            model=WeatherPredictionModel(name=model, abbrev=model),
+                            model_run_timestamp=model_run_timestamp,
+                            prediction_timestamps=[prediction_timestamp, ])
```
Comment (andrea-williams) [inline api/app/c_haines/fetch.py:158]:
```suggestion
                            prediction_timestamps=[prediction_timestamp])
```

File: api/app/db/crud/c_haines.py [hunk 3]
```diff
@@ -123,13 +112,13 @@ def get_prediction_geojson(session: Session,
...
                            prediction_timestamp: datetime):
-    """ Get the geojson for a particular prediction """
+    """ Get the geojson for a particular prediction, in WGS84 """
     query = text("""select json_build_object(
         'type', 'FeatureCollection',
         'features', json_agg(ST_AsGeoJSON(t.*)::json)
     )
         from (
-        select geom, c_haines_index from c_haines_polygons
+        select ST_Transform(geom, 'epsg:4269', 'epsg:4326'), c_haines_index from c_haines_polygons
```
Comment (andrea-williams) [inline api/app/db/crud/c_haines.py:121]: what transformation is going on here? Why is the coordinate system being changed?

Comment (andrea-williams) [reply api/app/db/crud/c_haines.py:121]: Nvm - found the answer to my question in another file. But I'd recommend adding a comment here to explain why the coordinate system needs to be transformed.

File: api/app/weather_models/process_grib.py [hunk 4]
```diff
@@ -35,11 +36,11 @@ class ModelRunInfo():
     """
 
     def __init__(self):
-        self.model_enum: ModelEnum = None
-        self.projection: ProjectionEnum = None
-        self.model_run_timestamp = None
-        self.prediction_timestamp = None
-        self.variable_name = None
+        self.model_enum: Union(ModelEnum, None) = None
```
Comment (conbrad) [inline api/app/weather_models/process_grib.py:39]: `Optional[ModelEnum] = None`
`Optional[ProjectionEnum] = None`
`Optional[datetime] = None`
`Optional[str] = None`

**Review by conbrad — APPROVED** [review]
LGTM just minor stuff

File: api/app/weather_models/process_grib.py (hunk 4 above)
Comment (andrea-williams) [inline api/app/weather_models/process_grib.py]: can't these Unions be replaced with `Optional`?

File: api/app/c_haines/migrate_kml.py [hunk 5]
```diff
@@ -6,46 +6,39 @@
...
-from app.c_haines.severity_index import generate_full_kml_path
+from app.c_haines.object_store import ObjectTypeEnum, generate_full_object_store_path
 
 logger = logging.getLogger(__name__)
 
 
 class KMLGeojsonPolygonIterator:
     """ Generator that produces a kml polygon for every geojson feature. This generator assumes GeoJSON
-    as provided by DB query. """
+    as provided by DB query. No projection transformation is required, as the geojson from the database
```
Comment (andrea-williams) [inline api/app/c_haines/migrate_kml.py:29]:
```suggestion
    is provided by DB query. No projection transformation is required, as the geojson from the database
```

File: api/app/weather_models/process_grib.py (hunk 4 above)
Comment (Sybrand) [reply api/app/weather_models/process_grib.py:39]: Thanks - missed that one.

File: api/app/c_haines/migrate_kml.py [hunk 6]
```diff
@@ -83,24 +77,26 @@ def main():
...
+            with io.StringIO() as kml_file:
+                # iterate through the parts of the kml generator, writing each part to file.
+                for part in generate_kml_prediction(polygon_iterator,
+                                                    prediction_model,
+                                                    model_run.model_run_timestamp,
+                                                    prediction.prediction_timestamp):
+                    kml_file.write(part)
+                # smash it into binary
+                kml_file.seek(0)
+                bio = io.BytesIO(kml_file.read().encode('utf8'))
```
[the remark this answers is not in the document]
Comment (Sybrand) [reply api/app/c_haines/migrate_kml.py:92]: Hope so! (Also an SSD can only handle so many writes before it dies!)

File: api/app/tests/c_haines/test_c_haines_worker.py [hunk 7]
```diff
@@ -63,7 +33,34 @@ def mock_requests_get(*args, **kwargs):
...
+            """ Mock minio object """
+
+            def list_objects(self, bucket_name, prefix=None, recursive=False,
+                             start_after=None, include_user_meta=False,
+                             include_version=False, use_api_v1=False) -> Iterator[Object]:
+                """ mock list objects.
+                """
+                if prefix in ('c-haines-polygons/kml/GDPS/2020/5/21/0/2020-05-21T00:00:00.kml',
+                              'c-haines-polygons/json/GDPS/2021/5/21/0/2021-05-21T00:00:00.json'):
+                    # The worker checks if the objects already exist - we want to of those checks
```
Comment (andrea-williams) [inline api/app/tests/c_haines/test_c_haines_worker.py:51]: confused by this comment - we want to do what?

File: api/app/utils/s3.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+"""
+from typing import Generator, Tuple
+from contextlib import asynccontextmanager
+from minio import Minio
+from aiobotocore.client import AioBaseClient
+from aiobotocore import get_session
+from app import config
+
+
+def get_minio_client() -> Tuple[Minio, str]:
```
Comment (andrea-williams) [inline api/app/utils/s3.py:11]: is this function still needed?

**Review by andrea-williams — APPROVED** [review]
There's a few areas where I'd like to see some more in-line documentation to explain what's going on, but nothing so vital that it's worth blocking deployment over, and otherwise all good.

File: api/app/db/crud/c_haines.py (hunk 3 above)
Comment (Sybrand) [reply api/app/db/crud/c_haines.py:121]: I here you - but this code is about to get deleted, so not going to bother. Answer is - in the DB it's NAD83 - but KML and GeoJSON are WGS84.

File: api/app/c_haines/fetch.py [hunk 9]
```diff
@@ -45,33 +35,35 @@ def fetch_model_run_kml_streamer(model: ModelEnum, model_run_timestamp: datetime
...
-        yield '</Link>\n'
-        yield '</NetworkLink>\n'
-
-    yield '</Folder>'
-    yield '</Folder>'
-    # Close the KML document.
-    yield '</Document>\n'
-    yield '</kml>\n'
-    logger.info('kml complete')
+    async with get_client() as (client, bucket):
```
[the remark this answers is not in the document]
Comment (Sybrand) [reply api/app/c_haines/fetch.py:38]: Two folders? Who needs that! Deleted, and added lots of comments for you.

File: api/app/c_haines/fetch.py (hunk 1 above)
Comment (Sybrand) [reply api/app/c_haines/fetch.py:104]: Ever so slightly different. The on is doing a model run, the other is doing a prediction. I agree they look very similar, but I'd prefer them not to share too much code.

File: api/app/tests/c_haines/test_c_haines_worker.py (hunk 7 above)
Comment (Sybrand) [reply api/app/tests/c_haines/test_c_haines_worker.py]: I've tried to make it a bit more clear (we want it to not find these two files, but find all others).

Comment (andrea-williams) [reply api/app/tests/c_haines/test_c_haines_worker.py]: ohhhhh wait I just got struck with a bolt of clarity - see suggested change

Comment (andrea-williams) [inline api/app/tests/c_haines/test_c_haines_worker.py]:
```suggestion
                    # The worker checks if the objects already exist - we want two of those checks
```

**Pull request merged**